File: shared/NeuralBot.cs
using Robocode.TankRoyale.BotApi;
using Robocode.TankRoyale.BotApi.Graphics;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

public class BulletInfo
{
    public double X { get; set; }
    public double Y { get; set; }
    public double Direction { get; set; }
    public double Power { get; set; }
    public bool IsEnemy { get; set; }
    public int Tick { get; set; }
}

public class NeuralBot : Bot
{
    private InferenceSession session;
    private List<BulletInfo> activeBullets = new();
    private float[] lastEnemyData = new float[5]; // x, y, dir, energy, distance

    public NeuralBot()
    {
        var botInfo = BotInfo.FromFile("NeuralBot.json");
        session = new InferenceSession("model.onnx");
    }

    public override void Run()
    {
        // Load colors
        var colorsJson = File.ReadAllText("colors.json");
        var colors = JsonSerializer.Deserialize<Dictionary<string, int[]>>(colorsJson);
        BodyColor = Color.FromRgb(colors["bodyColor"][0], colors["bodyColor"][1], colors["bodyColor"][2]);
        TurretColor = Color.FromRgb(colors["turretColor"][0], colors["turretColor"][1], colors["turretColor"][2]);
        RadarColor = Color.FromRgb(colors["radarColor"][0], colors["radarColor"][1], colors["radarColor"][2]);
        BulletColor = Color.FromRgb(colors["bulletColor"][0], colors["bulletColor"][1], colors["bulletColor"][2]);
        ScanColor = Color.FromRgb(colors["scanColor"][0], colors["scanColor"][1], colors["scanColor"][2]);

        while (IsRunning)
        {
            // Get current state
            var state = GetStateVector();

            // Infer actions
            var actions = InferActions(state);

            // Execute actions
            ExecuteActions(actions);

            // Scan
            SetTurnRadarRight(360);
            Go();
        }
    }

    private float[] GetStateVector()
    {
        // Normalize inputs
        float normX = (float)X / ArenaWidth;
        float normY = (float)Y / ArenaHeight;
        float normDir = (float)Direction / 360f;
        float normEnergy = (float)Energy / 100f;
        float normGunHeat = GunHeat / 3f;
        float normGunDir = (float)GunDirection / 360f;
        float normRadarDir = (float)RadarDirection / 360f;
        float normArenaW = 1f;
        float normArenaH = 1f;
        float normEnemyX = lastEnemyData[0] / ArenaWidth;
        float normEnemyY = lastEnemyData[1] / ArenaHeight;
        float normEnemyDir = lastEnemyData[2] / 360f;
        float normEnemyEnergy = lastEnemyData[3] / 100f;
        float normDistance = lastEnemyData[4] / Math.Max(ArenaWidth, ArenaHeight);

        var state = new List<float> { normX, normY, normDir, normEnergy, normGunHeat, normGunDir, normRadarDir, normArenaW, normArenaH, normEnemyX, normEnemyY, normEnemyDir, normEnemyEnergy, normDistance };

        // Add up to 5 closest bullets
        var enemyBullets = activeBullets.Where(b => b.IsEnemy).ToList();
        var sortedBullets = enemyBullets.OrderBy(b => DistanceTo(b.X, b.Y)).Take(5);
        foreach (var bullet in sortedBullets)
        {
            float relAngle = (float)((Math.Atan2(bullet.Y - Y, bullet.X - X) * 180 / Math.PI - Direction + 360) % 360) / 360f;
            float dist = (float)DistanceTo(bullet.X, bullet.Y) / Math.Max(ArenaWidth, ArenaHeight);
            float isEnemy = bullet.IsEnemy ? 1f : 0f;
            state.Add(relAngle);
            state.Add(dist);
            state.Add(isEnemy);
        }
        // Pad to 5 bullets
        while (state.Count < 12 + 15) state.Add(0f);

        return state.ToArray();
    }

    private float[] InferActions(float[] state)
    {
        var inputTensor = new DenseTensor<float>(state, new int[] { 1, 12 });
        var inputs = new List<NamedOnnxValue> { NamedOnnxValue.CreateFromTensor("input", inputTensor) };
        using var results = session.Run(inputs);
        var outputTensor = results.First().AsTensor<float>();
        return outputTensor.ToArray(); // 4 values: body_turn, move_dist, gun_turn, fire_power
    }

    private void ExecuteActions(float[] actions)
    {
        float bodyTurn = actions[0] * 180f; // -180 to 180
        float moveDist = actions[1] * 100f; // -100 to 100
        float gunTurn = actions[2] * 180f; // -180 to 180
        float firePower = Math.Clamp(actions[3] * 3f, 0f, 3f); // 0 to 3

        SetTurnLeft(bodyTurn);
        if (moveDist > 0)
            Forward(moveDist);
        else
            Back(-moveDist);
        SetTurnGunLeft(gunTurn);
        if (firePower > 0 && GunHeat == 0)
            Fire(firePower);
    }

    public override void OnScannedBot(ScannedBotEvent evt)
    {
        lastEnemyData[0] = (float)evt.X;
        lastEnemyData[1] = (float)evt.Y;
        lastEnemyData[2] = (float)evt.Direction;
        lastEnemyData[3] = (float)evt.Energy;
        lastEnemyData[4] = (float)DistanceTo(evt.X, evt.Y);
    }

    public override void OnBulletFired(BulletFiredEvent evt)
    {
        activeBullets.Add(new BulletInfo
        {
            X = evt.Bullet.X,
            Y = evt.Bullet.Y,
            Direction = evt.Bullet.Direction,
            Power = evt.Bullet.Power,
            IsEnemy = evt.Bullet.OwnerId != Id,
            Tick = Time
        });
    }

    public override void OnBulletHit(BulletHitEvent evt)
    {
        activeBullets.RemoveAll(b => b.X == evt.Bullet.X && b.Y == evt.Bullet.Y);
    }

    public override void OnBulletHitBullet(BulletHitBulletEvent evt)
    {
        activeBullets.RemoveAll(b => b.X == evt.Bullet.X && b.Y == evt.Bullet.Y);
        activeBullets.RemoveAll(b => b.X == evt.HitBullet.X && b.Y == evt.HitBullet.Y);
    }

    public override void OnHitByBullet(HitByBulletEvent evt)
    {
        // Optional: adjust behavior
    }

    public override void OnHitWall(HitWallEvent evt)
    {
        // Optional: adjust behavior
    }
}

class Program
{
    static void Main() => new NeuralBot().Run();
}

File: shared/NeuralBot.py
import json
import math
import statistics
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List

import numpy as np
import onnxruntime as ort
from robocode_tank_royale.bot_api import Bot, BotInfo
from robocode_tank_royale.bot_api.events import (
    BulletFiredEvent,
    BulletHitBulletEvent,
    BulletHitBotEvent,
    HitByBulletEvent,
    HitWallEvent,
    ScannedBotEvent,
)
from robocode_tank_royale.bot_api.graphics import Color


@dataclass
class BulletInfo:
    x: float
    y: float
    direction: float
    power: float
    is_enemy: bool
    tick: int


class NeuralBot(Bot):
    def __init__(self) -> None:
        super().__init__(BotInfo.from_file("NeuralBot.json"))
        self.session = ort.InferenceSession("model.onnx")
        self.active_bullets: Dict[int, BulletInfo] = {}
        self.last_enemy_data: List[float] = [
            0.0
        ] * 6  # x, y, dir, energy, distance, speed
        self.input_tensor_info: Dict[str, bool] = {
            "includeEnemy": True,
            "includeBullets": False,
            "includeRadar": False,
        }
        self.expected_input_size = self._get_expected_input_size()
        self.inference_times: List[float] = []
        self.skipped_turn_count = 0

    def run(self) -> None:
        config = self._load_config()
        self.input_tensor_info = {
            "includeEnemy": bool(
                config.get("inputTensorInfo", {}).get("includeEnemy", True)
            ),
            "includeBullets": bool(
                config.get("inputTensorInfo", {}).get("includeBullets", False)
            ),
            "includeRadar": bool(
                config.get("inputTensorInfo", {}).get("includeRadar", False)
            ),
        }

        self._apply_colors(config)

        input_info = config.get("inputInfo", {})
        input_type = str(input_info.get("inputType", "neural-network"))
        if input_type != "neural-network":
            print("Skipping neural network-specific steps.")
            return

        while self.running:
            state = self._get_state_vector()

            # Time the neural network inference
            inference_start = time.perf_counter()
            actions = self._infer_actions(state)
            inference_elapsed = (
                time.perf_counter() - inference_start
            ) * 1000  # Convert to ms
            self.inference_times.append(inference_elapsed)

            self._execute_actions(actions)
            self.set_turn_radar_right(360)
            self.go()

        self._print_stats()

    def _print_stats(self) -> None:
        """Called when the game ends. Print timing statistics."""
        if len(self.inference_times) == 0:
            return

        # Calculate statistics
        min_time = min(self.inference_times)
        max_time = max(self.inference_times)
        mean_time = statistics.mean(self.inference_times)
        median_time = statistics.median(self.inference_times)

        # Calculate percentiles
        p95_time = None
        p99_time = None
        try:
            p95_time = statistics.quantiles(self.inference_times, n=20)[
                18
            ]  # 95th percentile
            p99_time = statistics.quantiles(self.inference_times, n=100)[
                98
            ]  # 99th percentile
        except (statistics.StatisticsError, IndexError):
            pass  # Not enough data points for percentiles

        # Count slow inferences
        slow_threshold_ms = 1000  # 1 second
        slow_inferences = sum(1 for t in self.inference_times if t > slow_threshold_ms)

        # Print summary
        print(
            f"\r=== Timing Summary ===\r"
            + f"Total inferences: {len(self.inference_times)}\r"
            + f"Skipped turns: {self.skipped_turn_count}\r"
            + f"Inference time (ms):\r"
            + f"  Min:    {min_time:.2f}\r"
            + f"  Max:    {max_time:.2f}\r"
            + f"  Mean:   {mean_time:.2f}\r"
            + f"  Median: {median_time:.2f}\r"
            + (f"  P95:    {p95_time:.2f}\r" if p95_time is not None else "")
            + (f"  P99:    {p99_time:.2f}\r" if p99_time is not None else "")
            + f"Slow inferences (>{slow_threshold_ms}ms): {slow_inferences}\r"
            + f"==================================\r",
            flush=True,
        )

    def _load_config(self) -> dict:
        config_path = Path("config.json")
        if not config_path.exists():
            return {
                "inputTensorInfo": self.input_tensor_info,
                "colors": {},
                "inputInfo": {"inputType": "neural-network"},
            }

        with config_path.open("r", encoding="utf-8") as f:
            config = json.load(f)

        if "inputInfo" not in config:
            config["inputInfo"] = {"inputType": "neural-network"}

        if "inputTensorInfo" not in config:
            config["inputTensorInfo"] = self.input_tensor_info

        return config

    def _apply_colors(self, config: dict) -> None:
        colors = config.get("colors", {})

        def rgb(key: str, fallback: List[int]) -> Color:
            value = colors.get(key, fallback)
            if not isinstance(value, list) or len(value) != 3:
                value = fallback
            return Color.from_rgb(int(value[0]), int(value[1]), int(value[2]))

        self.body_color = rgb("bodyColor", [80, 120, 200])
        self.turret_color = rgb("turretColor", [80, 120, 200])
        self.radar_color = rgb("radarColor", [120, 180, 255])
        self.bullet_color = rgb("bulletColor", [255, 200, 80])
        if "scanColor" in colors:
            self.scan_color = rgb("scanColor", [140, 220, 140])

    def _get_state_vector(self) -> np.ndarray:
        arena_width = max(float(self.arena_width), 1.0)
        arena_height = max(float(self.arena_height), 1.0)
        arena_scale = max(arena_width, arena_height)

        state: List[float] = [
            float(self.x) / arena_width,
            float(self.y) / arena_height,
            float(self.direction) / 360.0,
            float(self.energy) / 100.0,
            float(self.gun_heat) / 3.0,
            float(self.gun_direction) / 360.0,
            float(self.radar_direction) / 360.0,
            1.0,
            1.0,
        ]

        if self.input_tensor_info.get("includeEnemy", True):
            state.extend(
                [
                    self.last_enemy_data[0] / arena_width,
                    self.last_enemy_data[1] / arena_height,
                    self.last_enemy_data[2] / 360.0,
                    self.last_enemy_data[3] / 100.0,
                    self.last_enemy_data[4] / arena_scale,
                    self.last_enemy_data[5] / 8.0,
                ]
            )

        if self.input_tensor_info.get("includeBullets", False):
            enemy_bullets = sorted(
                (b for b in self.active_bullets.values() if b.is_enemy),
                key=lambda b: self.distance_to(b.x, b.y),
            )[:5]

            for bullet in enemy_bullets:
                rel_angle = (
                    (
                        math.degrees(math.atan2(bullet.y - self.y, bullet.x - self.x))
                        - self.direction
                        + 360.0
                    )
                    % 360.0
                ) / 360.0
                dist = float(self.distance_to(bullet.x, bullet.y)) / arena_scale
                state.extend([rel_angle, dist, 1.0])

            while len(state) < 27:
                state.append(0.0)

        return np.asarray(state, dtype=np.float32)

    def _infer_actions(self, state: np.ndarray) -> np.ndarray:
        if self.expected_input_size > 0:
            if state.size > self.expected_input_size:
                state = state[: self.expected_input_size]
            elif state.size < self.expected_input_size:
                pad = np.zeros(self.expected_input_size - state.size, dtype=np.float32)
                state = np.concatenate((state, pad), axis=0)

        input_name = self.session.get_inputs()[0].name
        result = self.session.run(None, {input_name: state.reshape(1, -1)})
        output = np.asarray(result[0], dtype=np.float32).reshape(-1)
        if output.size < 4:
            padded = np.zeros(4, dtype=np.float32)
            padded[: output.size] = output
            return padded
        return output[:4]

    def _get_expected_input_size(self) -> int:
        shape = self.session.get_inputs()[0].shape
        if len(shape) < 2:
            return 0
        size = shape[1]
        return int(size) if isinstance(size, int) else 0

    def _execute_actions(self, actions: np.ndarray) -> None:
        body_turn = float(actions[0]) * 180.0
        move_dist = float(actions[1]) * 100.0
        gun_turn = float(actions[2]) * 180.0
        fire_power = max(0.0, min(float(actions[3]) * 3.0, 3.0))

        self.set_turn_left(body_turn)
        if move_dist > 0:
            self.set_forward(move_dist)
        elif move_dist < 0:
            self.set_back(-move_dist)

        self.set_turn_gun_left(gun_turn)
        if fire_power > 0.0 and self.gun_heat <= 0.0:
            self.set_fire(fire_power)

    def on_scanned_bot(self, evt: ScannedBotEvent) -> None:
        self.last_enemy_data[0] = float(evt.x)
        self.last_enemy_data[1] = float(evt.y)
        self.last_enemy_data[2] = float(evt.direction)
        self.last_enemy_data[3] = float(evt.energy)
        self.last_enemy_data[4] = float(self.distance_to(evt.x, evt.y))
        self.last_enemy_data[5] = float(evt.speed)

    def on_bullet_fired(self, evt: BulletFiredEvent) -> None:
        bullet = evt.bullet
        self.active_bullets[int(bullet.bullet_id)] = BulletInfo(
            x=float(bullet.x),
            y=float(bullet.y),
            direction=float(bullet.direction),
            power=float(bullet.power),
            is_enemy=int(bullet.owner_id) != int(self.my_id),
            tick=int(self.turn_number),
        )

    def on_bullet_hit(self, evt: BulletHitBotEvent) -> None:
        self.active_bullets.pop(int(evt.bullet.bullet_id), None)

    def on_bullet_hit_bullet(self, evt: BulletHitBulletEvent) -> None:
        self.active_bullets.pop(int(evt.bullet.bullet_id), None)
        self.active_bullets.pop(int(evt.hit_bullet.bullet_id), None)

    def on_hit_by_bullet(self, evt: HitByBulletEvent) -> None:
        pass

    def on_hit_wall(self, evt: HitWallEvent) -> None:
        pass

    def on_skipped_turn(self, evt) -> None:
        """Called when a turn is skipped due to timeout or other server-side issues."""
        self.skipped_turn_count += 1

    def on_game_ended(self, evt) -> None:
        """Called when the game ends. Print timing statistics."""
        # self._print_stats()
        pass


if __name__ == "__main__":
    NeuralBot().start()
